File: packages/edge/src/services/tokenMirrorService.ts
import axios from "axios";
import Redis from "ioredis";
import { config } from "../config/env";
import logger from "../config/logger";

/**
 * Token structure mirrored from backend
 * Note: id is number to match backend database ID for usage tracking
 */
export interface MirroredToken {
  id: number;
  tokenName: string;
  tokenValue: string;
  tokenType: "client" | "server" | "edge" | "all";
  allowAllEnvironments: boolean;
  environments: string[]; // ['*'] for all, or list of environment names
  expiresAt: string | null;
  createdAt: string;
  updatedAt: string;
}

/**
 * Token validation result
 */
export interface TokenValidationResult {
  valid: boolean;
  token?: MirroredToken;
  reason?: "not_found" | "expired" | "invalid_type" | "invalid_environment";
}

/**
 * TokenMirrorService - Mirrors all API tokens from backend to Edge memory
 * Uses ioredis for real-time token change notifications via Redis PubSub
 */
class TokenMirrorService {
  private tokens: Map<string, MirroredToken> = new Map(); // tokenValue -> token
  private tokenById: Map<number, MirroredToken> = new Map(); // id -> token
  private subscriber: Redis | null = null;
  private initialized = false;
  private readonly CHANNEL_NAME = "gatrix-sdk-events";

  /**
   * Initialize the token mirror service
   * Fetches all tokens from backend and subscribes to change events
   * Note: This should be called after SDK initialization (backend is ready)
   */
  async initialize(): Promise<void> {
    if (this.initialized) {
      logger.warn("[TokenMirror] Already initialized");
      return;
    }

    logger.info("[TokenMirror] Initializing token mirror service...");

    // Fetch initial tokens
    await this.fetchAllTokens();

    // Subscribe to token change events
    await this.subscribeToEvents();

    this.initialized = true;
    logger.info(`[TokenMirror] Initialized with ${this.tokens.size} tokens`);
  }

  /**
   * Shutdown the token mirror service
   */
  async shutdown(): Promise<void> {
    if (this.subscriber) {
      try {
        await this.subscriber.unsubscribe(this.CHANNEL_NAME);
        await this.subscriber.quit();
        this.subscriber = null;
      } catch (error) {
        logger.error("[TokenMirror] Error during shutdown:", error);
      }
    }
    this.tokens.clear();
    this.tokenById.clear();
    this.initialized = false;
    logger.info("[TokenMirror] Shutdown complete");
  }

  /**
   * Fetch all tokens from backend
   */
  async fetchAllTokens(): Promise<void> {
    try {
      const response = await axios.get(
        `${config.gatrixUrl}/api/v1/server/internal/tokens`,
        {
          headers: {
            "x-api-token": config.apiToken,
            "x-application-name": config.applicationName,
          },
          timeout: 10000,
        },
      );

      if (response.data?.success && response.data?.data?.tokens) {
        const tokens: MirroredToken[] = response.data.data.tokens;

        // Clear existing tokens
        this.tokens.clear();
        this.tokenById.clear();

        // Add new tokens
        for (const token of tokens) {
          this.tokens.set(token.tokenValue, token);
          this.tokenById.set(token.id, token);
        }

        logger.info(
          `[TokenMirror] Fetched ${tokens.length} tokens from backend`,
        );
      } else {
        logger.error(
          "[TokenMirror] Invalid response from backend:",
          response.data,
        );
      }
    } catch (error: any) {
      logger.error("[TokenMirror] Failed to fetch tokens:", error.message);
      throw error;
    }
  }

  /**
   * Subscribe to token change events via Redis PubSub
   */
  private async subscribeToEvents(): Promise<void> {
    try {
      this.subscriber = new Redis({
        host: config.redis.host,
        port: config.redis.port,
        password: config.redis.password || undefined,
        db: config.redis.db,
        lazyConnect: true,
      });

      await this.subscriber.connect();

      this.subscriber.on("message", (channel: string, message: string) => {
        if (channel === this.CHANNEL_NAME) {
          this.handleEvent(message);
        }
      });

      await this.subscriber.subscribe(this.CHANNEL_NAME);

      logger.info(
        `[TokenMirror] Subscribed to Redis channel: ${this.CHANNEL_NAME}`,
      );
    } catch (error: any) {
      logger.error(
        "[TokenMirror] Failed to subscribe to events:",
        error.message,
      );
      // Continue without real-time updates - will rely on manual refresh
    }
  }

  /**
   * Handle incoming event from Redis PubSub
   */
  private handleEvent(message: string): void {
    try {
      const event = JSON.parse(message);

      if (!event.type?.startsWith("api_token.")) {
        return; // Not a token event
      }

      logger.info(`[TokenMirror] Received event: ${event.type}`, {
        id: event.data?.id,
      });

      // For any token change, refetch all tokens
      // This is simpler and more reliable than incremental updates
      this.fetchAllTokens().catch((err) => {
        logger.error(
          "[TokenMirror] Failed to refetch tokens after event:",
          err.message,
        );
      });
    } catch (error: any) {
      logger.error("[TokenMirror] Failed to parse event:", error.message);
    }
  }

  /**
   * Validate a token
   */
  validateToken(
    tokenValue: string,
    requiredType: "client" | "server" | "all",
    environment?: string,
  ): TokenValidationResult {
    // Check for unsecured client token (for testing purposes, client -> edge)
    // Note: id=0 is used for unsecured tokens to skip usage tracking
    if (tokenValue === config.unsecuredClientToken) {
      logger.debug("Unsecured client token used for testing");
      const unsecuredToken: MirroredToken = {
        id: 0, // 0 indicates unsecured token, usage tracking will be skipped
        tokenName: "Unsecured Client Token (Testing)",
        tokenValue: config.unsecuredClientToken,
        tokenType: "all",
        allowAllEnvironments: true,
        environments: ["*"],
        expiresAt: null,
        createdAt: new Date().toISOString(),
        updatedAt: new Date().toISOString(),
      };
      return { valid: true, token: unsecuredToken };
    }

    const token = this.tokens.get(tokenValue);

    if (!token) {
      return { valid: false, reason: "not_found" };
    }

    // Check expiration
    if (token.expiresAt) {
      const expiresAt = new Date(token.expiresAt);
      if (expiresAt < new Date()) {
        return { valid: false, token, reason: "expired" };
      }
    }

    // Check token type
    // 'all' type can access both client and server APIs
    if (token.tokenType !== "all" && token.tokenType !== requiredType) {
      return { valid: false, token, reason: "invalid_type" };
    }

    // Check environment access
    if (environment && !token.allowAllEnvironments) {
      if (
        !token.environments.includes(environment) &&
        !token.environments.includes("*")
      ) {
        return { valid: false, token, reason: "invalid_environment" };
      }
    }

    return { valid: true, token };
  }

  /**
   * Get token by value
   */
  getToken(tokenValue: string): MirroredToken | undefined {
    return this.tokens.get(tokenValue);
  }

  /**
   * Get token by ID
   */
  getTokenById(id: number): MirroredToken | undefined {
    return this.tokenById.get(id);
  }

  /**
   * Get all tokens count
   */
  getTokenCount(): number {
    return this.tokens.size;
  }

  /**
   * Check if service is initialized
   */
  isInitialized(): boolean {
    return this.initialized;
  }

  /**
   * Force refresh tokens from backend
   */
  async refresh(): Promise<void> {
    await this.fetchAllTokens();
  }
}

export const tokenMirrorService = new TokenMirrorService();

File: packages/edge-dotnet/src/Gatrix.Edge/Services/TokenMirrorService.cs
using System.Collections.Concurrent;
using System.Net.Http.Json;
using System.Text.Json;
using Gatrix.Edge.Models;
using Gatrix.Edge.Options;
using Microsoft.Extensions.Options;
using StackExchange.Redis;

namespace Gatrix.Edge.Services;

/// <summary>
/// Mirrors all API tokens from backend to Edge memory.
/// Uses Redis Pub/Sub for real-time token change notifications.
/// </summary>
public class TokenMirrorService : IHostedService, IDisposable
{
    private readonly ConcurrentDictionary<string, MirroredToken> _tokensByValue = new();
    private readonly ConcurrentDictionary<int, MirroredToken> _tokensById = new();
    private readonly EdgeOptions _options;
    private readonly IHttpClientFactory _httpClientFactory;
    private readonly ILogger<TokenMirrorService> _logger;
    private ISubscriber? _subscriber;
    private IConnectionMultiplexer? _redis;
    private bool _initialized;
    private const string ChannelName = "gatrix-sdk-events";
    private readonly string _cacheFilePath;

    public TokenMirrorService(
        IOptions<EdgeOptions> options,
        IHttpClientFactory httpClientFactory,
        ILogger<TokenMirrorService> logger)
    {
        _options = options.Value;
        _httpClientFactory = httpClientFactory;
        _logger = logger;
        
        // Setup cache file path (matches SDK convention)
        var cacheDir = Path.Combine(Directory.GetCurrentDirectory(), ".gatrix_cache");
        if (!Directory.Exists(cacheDir)) Directory.CreateDirectory(cacheDir);
        _cacheFilePath = Path.Combine(cacheDir, "edge_tokens.json");
    }

    public async Task StartAsync(CancellationToken cancellationToken)
    {
        _logger.LogInformation("[TokenMirror] Initializing token mirror service...");

        // 1. Try to load from local cache first
        await LoadFromCacheAsync(cancellationToken);

        // 2. Fetch from backend in background
        _ = FetchAllTokensAsync(cancellationToken);

        await SubscribeToEventsAsync();

        _initialized = true;
        _logger.LogInformation("[TokenMirror] Initialized (local count: {Count})", _tokensByValue.Count);
    }

    private async Task LoadFromCacheAsync(CancellationToken ct)
    {
        try
        {
            if (File.Exists(_cacheFilePath))
            {
                var json = await File.ReadAllTextAsync(_cacheFilePath, ct);
                var tokens = JsonSerializer.Deserialize<List<MirroredToken>>(json,
                    new JsonSerializerOptions { PropertyNameCaseInsensitive = true });

                if (tokens != null)
                {
                    UpdateInternalMaps(tokens);
                    _logger.LogInformation("[TokenMirror] Loaded {Count} tokens from local cache", tokens.Count);
                }
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "[TokenMirror] Failed to load tokens from local cache");
        }
    }

    private async Task SaveToCacheAsync(List<MirroredToken> tokens, CancellationToken ct)
    {
        try
        {
            var json = JsonSerializer.Serialize(tokens, new JsonSerializerOptions { WriteIndented = true });
            await File.WriteAllTextAsync(_cacheFilePath, json, ct);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "[TokenMirror] Failed to save tokens to local cache");
        }
    }

    private void UpdateInternalMaps(List<MirroredToken> tokens)
    {
        _tokensByValue.Clear();
        _tokensById.Clear();

        foreach (var token in tokens)
        {
            _tokensByValue[token.TokenValue] = token;
            _tokensById[token.Id] = token;
        }
    }

    public async Task StopAsync(CancellationToken cancellationToken)
    {
        if (_subscriber != null)
        {
            await _subscriber.UnsubscribeAsync(RedisChannel.Literal(ChannelName));
        }
        if (_redis != null)
        {
            await _redis.CloseAsync();
            _redis.Dispose();
        }
        _tokensByValue.Clear();
        _tokensById.Clear();
        _initialized = false;
        _logger.LogInformation("[TokenMirror] Shutdown complete");
    }

    public async Task FetchAllTokensAsync(CancellationToken ct = default)
    {
        try
        {
            var client = _httpClientFactory.CreateClient("GatrixBackend");
            var response = await client.GetAsync("/api/v1/server/internal/tokens", ct);
            response.EnsureSuccessStatusCode();

            var json = await response.Content.ReadAsStringAsync(ct);
            var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            if (root.TryGetProperty("success", out var success) && success.GetBoolean() &&
                root.TryGetProperty("data", out var data) &&
                data.TryGetProperty("tokens", out var tokensElement))
            {
                var tokens = JsonSerializer.Deserialize<List<MirroredToken>>(
                    tokensElement.GetRawText(),
                    new JsonSerializerOptions { PropertyNameCaseInsensitive = true });

                if (tokens != null)
                {
                    UpdateInternalMaps(tokens);
                    _logger.LogInformation("[TokenMirror] Fetched {Count} tokens from backend", tokens.Count);

                    // Persist to local cache
                    await SaveToCacheAsync(tokens, ct);
                }
            }
            else
            {
                _logger.LogError("[TokenMirror] Invalid response from backend");
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "[TokenMirror] Failed to fetch tokens from backend. Using current cache (if any).");
            // Non-critical if we already have local cache
        }
    }

    private async Task SubscribeToEventsAsync()
    {
        try
        {
            var configOptions = new ConfigurationOptions
            {
                EndPoints = { { _options.Redis.Host, _options.Redis.Port } },
                DefaultDatabase = _options.Redis.Db,
                AbortOnConnectFail = false,
            };
            if (!string.IsNullOrEmpty(_options.Redis.Password))
            {
                configOptions.Password = _options.Redis.Password;
            }

            _redis = await ConnectionMultiplexer.ConnectAsync(configOptions);
            _subscriber = _redis.GetSubscriber();

            await _subscriber.SubscribeAsync(RedisChannel.Literal(ChannelName), (channel, message) =>
            {
                HandleEvent(message!);
            });

            _logger.LogInformation("[TokenMirror] Subscribed to Redis channel: {Channel}", ChannelName);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "[TokenMirror] Failed to subscribe to events");
            // Continue without real-time updates
        }
    }

    private void HandleEvent(string message)
    {
        try
        {
            var doc = JsonDocument.Parse(message);
            var root = doc.RootElement;

            if (!root.TryGetProperty("type", out var typeEl))
                return;

            var type = typeEl.GetString();
            if (type == null || !type.StartsWith("api_token."))
                return;

            _logger.LogInformation("[TokenMirror] Received event: {Type}", type);

            // Refetch all tokens on any token change
            _ = FetchAllTokensAsync();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "[TokenMirror] Failed to parse event");
        }
    }

    /// <summary>
    /// Validate a token locally.
    /// </summary>
    public TokenValidationResult ValidateToken(string tokenValue, string requiredType, string? environmentId = null)
    {
        // Check unsecured client token
        if (tokenValue == _options.UnsecuredClientToken)
        {
            _logger.LogDebug("Unsecured client token used for testing");
            return new TokenValidationResult
            {
                Valid = true,
                Token = new MirroredToken
                {
                    Id = 0,
                    TokenName = "Unsecured Client Token (Testing)",
                    TokenValue = _options.UnsecuredClientToken,
                    TokenType = "all",
                    AllowAllEnvironments = true,
                    Environments = new List<string> { "*" },
                    CreatedAt = DateTime.UtcNow.ToString("o"),
                    UpdatedAt = DateTime.UtcNow.ToString("o"),
                }
            };
        }

        if (!_tokensByValue.TryGetValue(tokenValue, out var token))
        {
            return new TokenValidationResult { Valid = false, Reason = "not_found" };
        }

        // Check expiration
        if (!string.IsNullOrEmpty(token.ExpiresAt))
        {
            if (DateTime.TryParse(token.ExpiresAt, out var expiresAt) && expiresAt < DateTime.UtcNow)
            {
                return new TokenValidationResult { Valid = false, Token = token, Reason = "expired" };
            }
        }

        // Check token type
        if (token.TokenType != "all" && token.TokenType != requiredType)
        {
            return new TokenValidationResult { Valid = false, Token = token, Reason = "invalid_type" };
        }

        // Check environment access
        if (environmentId != null && !token.AllowAllEnvironments)
        {
            if (!token.Environments.Contains(environmentId) && !token.Environments.Contains("*"))
            {
                return new TokenValidationResult { Valid = false, Token = token, Reason = "invalid_environment" };
            }
        }

        return new TokenValidationResult { Valid = true, Token = token };
    }

    public MirroredToken? GetToken(string tokenValue) =>
        _tokensByValue.TryGetValue(tokenValue, out var t) ? t : null;

    public MirroredToken? GetTokenById(int id) =>
        _tokensById.TryGetValue(id, out var t) ? t : null;

    public int GetTokenCount() => _tokensByValue.Count;
    public bool IsInitialized => _initialized;

    public void Dispose()
    {
        _redis?.Dispose();
        GC.SuppressFinalize(this);
    }
}
